gitlab-runner/src/lib.rs

```diff
@@ -106,6 +108,52 @@ macro_rules! outputln {
 pub type JobResult = Result<(), ()>;
 pub use client::Phase;
 
+/// A file that a [`JobHandler`] is willing to upload to the server
+///
+/// A [`JobHandler`] will be queried for the set of files it is able
+/// to upload to the server. These might be on disk, or they might be
+/// generated or downloaded from some other source on demand. The
+/// `get_path` method is required so that the globbing Gitlab expects
+/// can be performed without the handler needing to be involved.
+pub trait UploadableFile: Eq {
+    /// The type of the data stream returned by
+    /// [`get_data`](Self::get_data)
+    type Data<'a>: AsyncRead + Send + Unpin
+    where
+        Self: 'a;
+
+    /// Get the logical path of the file.
+    ///
+    /// This is the path on disk from the root of the checkout for
+    /// Gitlab's own runner. It should match the paths that users are
+    /// expected to specify when requesting artifacts in their jobs.
+    fn get_path(&self) -> Cow<'_, str>;
+
+    /// Get something that can provide the data for this file.
+    ///
+    /// This can be any implementor of [`AsyncRead`].
+    fn get_data(&self) -> Self::Data<'_>;
```

Comment:
So the reason I didn't do that was to avoid the `async_trait` mess, since it's not necessary; the magical future methods (i.e. `flatten_stream`) make this relatively convenient and simple to implement, once you have the recipe. I.e. you only really need one level of "asyncness", you can hide any amount of additional layers of async just by flattening, which simplified the interface, and avoids making another `async_trait`

Reply:
ah i didn't remember flatten_stream was a thing indeed :) 

I'm still a bit on the fence about this one because it means your Data type now has to be `FlattenStream<your data>`  and you'd probably do an async closure  or function to call flatten_stream on. Which seems a lot less pleasant for a user and given async_trait isn't introduced this for this avoiding the mess doesn't seem like a big benefit.  Oh also you'd have to be able to name the type so ; which you really can't do for functions or closures, which quickly forces things down the `Box<dyn> route`

Fwiw my hope is that in a year or two we'd be able to drop `async_trait` entirely and move to proper native async traits once Rust supports it; So i'm more treating `async_trait` as some annoying sprinklink you hav eto do for now to work around language limitations and ignoring that internally it does the same `Box<dyn>` tricks you're doing explicitly here :p